src/SessionManager/WebBrowser.cc

```diff
@@ -0,0 +1,135 @@
+        spdlog::debug("WebBrowser: using default browser.");
+        OpenSystemBrowser();
+    }
+}
+
+void WebBrowser::ParseCmd() {
+    if (_cmd[_cmd.size() - 1] == '&') {
+        _cmd.pop_back();
+    }
+    const std::string wildcard = "CARTA_URL";
+    auto wildcard_pos = _cmd.find(wildcard);
+    if (wildcard_pos != std::string::npos) {
+        _cmd = fmt::format("{0}{2} {1}", _cmd.substr(0, wildcard_pos), _cmd.substr(wildcard_pos + wildcard.size(), _cmd.size()), _url);
+    } else {
+        _cmd = fmt::format("{} {}", _cmd, _url);
+    }
+#if defined(__APPLE__)
+#else
+    std::istringstream isstream(_cmd);
+    std::copy(std::istream_iterator<std::string>(isstream), std::istream_iterator<std::string>(), std::back_inserter(_args));
+    fs::path path(_args[0]);
+    if (!fs::exists(path)) {
+        path = search_path(_args[0]);
+    }
+
+    if (path.empty()) {
+        spdlog::debug("Can't find {} in PATH, please check.", _args[0]);
+    } else {
+        _path_exists = true;
+    }
+    _args[0] = path.string();
+#endif
+}
+
+void WebBrowser::OpenSystemBrowser() {
+#if defined(__APPLE__)
+    std::string cmd = "open";
+#else
+    std::string cmd = "xdg-open";
+#endif
+    cmd = fmt::format("{} \"{}\"", cmd, _url).c_str();
+    spdlog::debug("WebBrowser: Trying to launch CARTA with the default browser using: {}", cmd);
+    const int ans = std::system(cmd.c_str());
+    if (ans) {
+        _status = false;
+        _error = "WebBrowser: Failed to open the default browser automatically.";
+    }
+}
+
+void WebBrowser::OpenBrowser() {
+    spdlog::debug("WebBrowser: Trying to launch CARTA with user provided browser command: {}", _cmd);
+#if defined(__APPLE__)
+    std::string cmd = "open -a " + _cmd;
+    const int ans = std::system(cmd.c_str());
+    if (ans) {
+        _status = false;
+        _error = "WebBrowser: Failed to open the browser automatically.";
+    }
+#else
+    pid_t pid = fork();
+    if (pid == 0) {
+        setsid();
+        // double fork
+        pid_t pid2 = fork();
+        signal(SIGHUP, SIG_IGN);
+        signal(SIGPIPE, SIG_IGN);
+        if (pid2 == 0) {
+            std::vector<char*> args;
+            for (auto& arg : _args) {
+                args.push_back(const_cast<char*>(arg.c_str()));
+            }
+            args.push_back(nullptr);
+
+            // redirect output using unistd
+            int fd = open("/dev/null", O_WRONLY);
+            dup2(fd, STDOUT_FILENO);
+            dup2(fd, STDERR_FILENO);
+            close(fd);
+
+            auto result = execv(args[0], args.data());
+            if (result == -1) {
+                spdlog::debug("WebBrowser: execv failed. CARTA can't start with the requiered settings in --browser.", result);
+            }
+            _exit(1);
+        } else if (pid2 == -1) {
+            spdlog::debug("WebBrowser: Failed to fork a new process. CARTA can't start with the requiered settings in --browser.");
+            if (chdir("/") == -1) {
```

Comment:
why do we need to `chdir` to `/`?

Reply:
I am sorry, I left this one without a reply. I believe this is mostly a good practice. I found this [article](http://www.microhowto.info/howto/cause_a_process_to_become_a_daemon.html) pretty good explaining how to start a daemon process. So I am just going to quote it here for completeness:

> If a daemon were to leave its current working directory unchanged then this would prevent the filesystem containing that directory from being unmounted while the daemon was running. It is therefore good practice for daemons to change their working directory to a safe location. This could be:
>- the root directory, or
>- a directory containing files that are needed by the daemon.
>(The latter option turns what would have been a drawback into a feature. If files are needed by the daemon then the filesystem containing them presumably should not be unmounted while the daemon is running. Moving the current working directory to that filesystem prevents this from happening.)

So, I believe this isn't a problem for any running browser, and we could leave the working path unchanged. But thinking further, if more low-level settings are passed to the browser in execution, it might be a good idea to get rid of potential problems.